File: tang/src/test/java/com/microsoft/inject/TweetExample.java
package com.microsoft.inject;

import static org.junit.Assert.*;

import javax.inject.Inject;
import javax.inject.Named;

import org.junit.After;
import org.junit.AfterClass;
import org.junit.Before;
import org.junit.BeforeClass;
import org.junit.Test;

public class TweetExample {
  private static interface TweetFactory {
    public String getTweet();
  }

  private static interface SMS {
    public void sendSMS(String msg, long phoneNumber);
  }

  private static class MockTweetFactory implements TweetFactory {
    @Inject
    public MockTweetFactory() {
    }

    @Override
    public String getTweet() {
      return "@tw #bbq bbqftw!!! gopher://vuwed.wefd/bbqftw!";
    }
  }

  private static class MockSMS implements SMS {
    @Inject
    public MockSMS() {
    }

    @Override
    public void sendSMS(String msg, long phoneNumber) {
      if (phoneNumber != 867 - 5309) {
        throw new IllegalArgumentException("Unknown recipient");
      }
      // success!
    }
  }

  private static class Tweeter {
    final TweetFactory tw;
    final SMS sms;
    final long phoneNumber;

    @Inject
    public Tweeter(TweetFactory tw, SMS sms,
        @Named("phone-number") long phoneNumber) {
      this.tw = tw;
      this.sms = sms;
      this.phoneNumber = phoneNumber;
    }

    public void sendMessage() {
      sms.sendSMS(tw.getTweet(), phoneNumber);
    }
  }

  private Namespace ns;

  @BeforeClass
  public static void setUpBeforeClass() throws Exception {
  }

  @AfterClass
  public static void tearDownAfterClass() throws Exception {
  }

  @Before
  public void setUp() throws Exception {
    ns = new Namespace();
    ns.registerClass(MockTweetFactory.class);
    ns.registerClass(MockSMS.class);
    ns.registerClass(Tweeter.class);
    // Pull in unknown dependencies (eg: TweetFactory) automagically.
    ns.resolveAllClasses();
  }

  @After
  public void tearDown() throws Exception {
    ns = null;
  }

  @Test
  public void test() throws Exception {
    Tang t = new Tang(ns);
    t.setDefaultImpl(TweetFactory.class, MockTweetFactory.class);
    t.setDefaultImpl(SMS.class, MockSMS.class);
    t.setNamedParameter(Tweeter.class.getName() + ".phone-number", 867 - 5309);
    Tweeter tw = (Tweeter) t.getInstance(Tweeter.class);
    tw.sendMessage();
  }

}

File: lang/cs/Org.Apache.REEF.Tang.Examples/TweetExample.cs
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

using System;
using Org.Apache.REEF.Tang.Annotations;

namespace Org.Apache.REEF.Tang.Examples
{
    public interface ISMS
    {
        void SendSMS(string msg, long phoneNumber);
    }

    public interface ITweetFactory
    {
        string GetTweet();
    } 

    public class MockTweetFactory : ITweetFactory 
    {
        [Inject]
        public MockTweetFactory() 
        {
        }

        public string GetTweet() 
        {
            return "@tw #bbq bbqftw!!! gopher://vuwed.wefd/bbqftw!";
        }
    }

    public class MockSMS : ISMS 
    {
        [Inject]
        public MockSMS() 
        {
        }

        public void SendSMS(string msg, long phoneNumber) 
        {
            if (phoneNumber != 8675309) 
            {
                throw new ArgumentException("Unknown recipient");
            }

            // success!
        }
    }

    public class Tweeter
    {
        readonly ITweetFactory tw;
        readonly ISMS sms;
        readonly long phoneNumber;

        [NamedParameter(Documentation = "Phone number", ShortName = "number", DefaultValue = "1800")]
        public class PhoneNumber : Name<long>
        { 
        }
        [Inject]
        public Tweeter(ITweetFactory tw, ISMS sms, [Parameter(typeof(PhoneNumber))] long phoneNumber)
        {
            this.tw = tw;
            this.sms = sms;
            this.phoneNumber = phoneNumber;
        }

        [Inject]
        public Tweeter([Parameter(typeof(PhoneNumber))] long phoneNumber)
        {
            this.phoneNumber = phoneNumber;
        }

        public void sendMessage()
        {
            sms.SendSMS(tw.GetTweet(), phoneNumber);
        }
    }
}
